File: backend/app/api/ws/ws_messages.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.room import PlayerOut


# Client → Server

class HostStartMsg(BaseModel):
    type: Literal["host_start"] = "host_start"


class HostNextMsg(BaseModel):
    type: Literal["host_next"] = "host_next"


class PlayerAnswerMsg(BaseModel):
    type: Literal["player_answer"] = "player_answer"
    option_idx: int = Field(..., ge=0, le=3)
    elapsed_ms: int = Field(..., ge=0)


# Server → Client

class PlayerJoinedMsg(BaseModel):
    type: Literal["player_joined"] = "player_joined"
    player: PlayerOut


class QuestionOptionView(BaseModel):
    """What the player SEES — WITHOUT `is_correct`."""
    text: str


class QuestionMsg(BaseModel):
    type: Literal["question"] = "question"
    index: int
    total: int
    text: str
    options: list[QuestionOptionView]
    time_limit_ms: int


class AnswerAckMsg(BaseModel):
    type: Literal["answer_ack"] = "answer_ack"
    is_correct: bool
    points_awarded: int
    total_score: int


class QuestionEndMsg(BaseModel):
    type: Literal["question_end"] = "question_end"
    correct_idx: int
    leaderboard: list[PlayerOut]


class GameEndMsg(BaseModel):
    type: Literal["game_end"] = "game_end"
    final_leaderboard: list[PlayerOut]


class ErrorMsg(BaseModel):
    type: Literal["error"] = "error"
    message: str

File: frontend/src/types/ws-messages.ts
import type { Player } from "./player";

export interface WsQuestionOption {
  text: string;
}

export interface PlayerJoinedMsg {
  type: "player_joined";
  player: Player;
}

export interface QuestionMsg {
  type: "question";
  index: number;
  total: number;
  text: string;
  options: WsQuestionOption[];
  time_limit_ms: number;
}

export interface AnswerAckMsg {
  type: "answer_ack";
  is_correct: boolean;
  points_awarded: number;
  total_score: number;
}

export interface QuestionEndMsg {
  type: "question_end";
  correct_idx: number;
  leaderboard: Player[];
}

export interface GameEndMsg {
  type: "game_end";
  final_leaderboard: Player[];
}

export interface ErrorMsg {
  type: "error";
  message: string;
}

export type ServerWsMessage =
  | PlayerJoinedMsg
  | QuestionMsg
  | AnswerAckMsg
  | QuestionEndMsg
  | GameEndMsg
  | ErrorMsg;

export interface HostStartMsg {
  type: "host_start";
}

export interface HostNextMsg {
  type: "host_next";
}

export interface PlayerAnswerMsg {
  type: "player_answer";
  option_idx: number;
  elapsed_ms: number;
}

export type ClientWsMessage = HostStartMsg | HostNextMsg | PlayerAnswerMsg;
